cpp/open3d/t/io/file_format/FileSPLAT.cpp

```diff
@@ -0,0 +1,354 @@
+                f_rest[i] = SH_C0;
+            }
+            std::memcpy(f_dc_ptr + index * 3, f_dc, 3 * sizeof(float));
+            std::memcpy(f_rest_ptr + index * 3, f_rest, 3 * sizeof(float));
+
+            // Calculate the opacity
+            float opacity[1];
+            if (color[3] == 0) {
+                opacity[0] = 0.0f;  // Handle division by zero
+            } else {
+                opacity[0] = -log(1 / (color[3] / 255.0f) - 1);
+            }
+            std::memcpy(opacity_ptr + index, opacity, sizeof(float));
+
+            float rot_float[4];
+            for (int i = 0; i < 4; i++) {
+                rot_float[i] = (rotation[i] / 128.0f) - 1.0f;
+            }
+            std::memcpy(rot_ptr + index * 4, rot_float, 4 * sizeof(float));
+
+            index++;
+            if (index % 1000 == 0) {
+                reporter.Update(file.CurPos());
+            }
+        }
+
+        // Report progress
+        reporter.Finish();
+        return true;
+    } catch (const std::exception &e) {
+        utility::LogWarning("Read SPLAT failed: {}", e.what());
+    }
+    return false;
+}
+
+bool WritePointCloudToSPLAT(const std::string &filename,
+                            const geometry::PointCloud &pointcloud,
+                            const open3d::io::WritePointCloudOption &params) {
+    FILE *splat_file = NULL;
+
+    // Validate Point Cloud
+    if (pointcloud.IsEmpty()) {
+        utility::LogWarning("Write SPLAT failed: point cloud has 0 points.");
+        return false;
+    }
+
+    // Validate Splat Data
+    geometry::TensorMap t_map(pointcloud.GetPointAttr().Contiguous());
+    if (!ValidSPLATData(pointcloud, t_map)) return false;
+
+    // Take pointers to attribute data
+    AttributePtr positions_attr = AttributePtr(
+            t_map["positions"].GetDtype(), t_map["positions"].GetDataPtr(), 3);
+    AttributePtr scale_attr = AttributePtr(t_map["scale"].GetDtype(),
+                                           t_map["scale"].GetDataPtr(), 3);
+
+    AttributePtr rot_attr =
+            AttributePtr(t_map["rot"].GetDtype(), t_map["rot"].GetDataPtr(), 4);
+
+    AttributePtr f_dc_attr = AttributePtr(t_map["f_dc"].GetDtype(),
+                                          t_map["f_dc"].GetDataPtr(), 3);
+
+    AttributePtr opacity_attr = AttributePtr(t_map["opacity"].GetDtype(),
+                                             t_map["opacity"].GetDataPtr(), 1);
+
+    // Total Gaussians
+    long num_gaussians =
+            static_cast<long>(pointcloud.GetPointPositions().GetLength());
+
+    // Open splat file
+    splat_file = fopen(filename.c_str(), "wb");
+    if (!splat_file) {
+        utility::LogWarning("Write SPLAT failed: unable to open file: {}.",
+                            filename);
+        return false;
+    }
+
+    // Write to SPLAT
+    utility::CountingProgressReporter reporter(params.update_progress);
+    reporter.SetTotal(num_gaussians);
+
```

Comment:
There's a sorting step before writing the `.splat` file to disk:
```py
    sorted_indices = np.argsort(
        -np.exp(vert["scale_0"] + vert["scale_1"] + vert["scale_2"])
        / (1 + np.exp(-vert["opacity"]))
    )
```
This re-orders the Gaussians so that "important" Gaussians get written (and also read while loading) earlier. This enables viewers to progressively update the view with more detail.

We can use the oneTBB [`parallel_sort`](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/v1.2-provisional-rev-1/elements/onetbb/source/algorithms/functions/parallel_sort_func) here for speed. See this [gist](https://gist.github.com/HViktorTsoi/58eabb4f7c5a303ced400bcfa816f6f5) for how to implement `argsort` with a C++ sort.

Reply:
Added sorting step.